src/aks-preview/azext_aks_preview/custom.py

```diff
@@ -684,6 +684,60 @@ def _add_monitoring_role_assignment(result, cluster_resource_id, cmd):
                        'assignment')
 
 
+def _add_ingress_appgw_addon_role_assignment(result, cmd):
+    service_principal_msi_id = None
+    # Check if service principal exists, if it does, assign permissions to service principal
+    # Else, provide permissions to MSI
+    if (
+            hasattr(result, 'service_principal_profile') and
+            hasattr(result.service_principal_profile, 'client_id') and
+            result.service_principal_profile.client_id != 'msi'
+    ):
+        service_principal_msi_id = result.service_principal_profile.client_id
+        is_service_principal = True
+    elif (
+            (hasattr(result, 'addon_profiles')) and
+            (CONST_INGRESS_APPGW_ADDON_NAME in result.addon_profiles) and
+            (hasattr(result.addon_profiles[CONST_INGRESS_APPGW_ADDON_NAME], 'identity')) and
+            (hasattr(result.addon_profiles[CONST_INGRESS_APPGW_ADDON_NAME].identity, 'object_id'))
+    ):
+        service_principal_msi_id = result.addon_profiles[CONST_INGRESS_APPGW_ADDON_NAME].identity.object_id
+        is_service_principal = False
```

Comment:
What if neither of condition is true?

And suggest to use `try` and `except` instead of multiple checks for the existence of attributes to follow [EAFP](https://docs.python.org/3/glossary.html#term-eafp)

Reply:
Better to put this complicate condition check into a function such as `is_msi` ?

Reply:
added `if service_principal_msi_id is not None:`